akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -23,23 +25,130 @@ trait CommandHandler[Command, Event, State] {
+   *
    * @param stateClass The handlers defined by this builder are used when the state is an instance of the `stateClass`
-   * @return A new, mutable, command handler builder
+   * @param statePredicate The handlers defined by this builder are used when the `statePredicate` is `true`
+   *
+   * @return A new, mutable, CommandHandlerBuilderByState
    */
-  def builder[Command, Event, S <: State, State](stateClass: Class[S]): CommandHandlerBuilder[Command, Event, S, State] =
-    new CommandHandlerBuilder(statePredicate = new Predicate[S] {
-      override def test(state: S): Boolean = state != null && stateClass.isAssignableFrom(state.getClass)
-    })
+  def forState[S <: State](stateClass: Class[S], statePredicate: Predicate[S]): CommandHandlerBuilderByState[Command, Event, S, State] = {
+    val builder = new CommandHandlerBuilderByState[Command, Event, S, State](stateClass, statePredicate)
+    builders = builder.asInstanceOf[CommandHandlerBuilderByState[Command, Event, State, State]] :: builders
+    builder
+  }
+
+  /**
+   * Use this method to define command handlers for a given subtype of your model. Useful when the model is defined as class hierarchy.
+   *
+   * @param stateClass The handlers defined by this builder are used when the state is an instance of the `stateClass`.
+   * @return A new, mutable, CommandHandlerBuilderByState
+   */
+  def forStateType[S <: State](stateClass: Class[S]): CommandHandlerBuilderByState[Command, Event, S, State] = {
+    val builder = CommandHandlerBuilderByState.builder[Command, Event, S, State](stateClass)
+    builders = builder.asInstanceOf[CommandHandlerBuilderByState[Command, Event, State, State]] :: builders
+    builder
+  }
+
+  /**
+   * The handlers defined by this builder are used when the state is `null`.
+   * This variant is particular useful when the empty state of your model is defined as `null`.
+   *
+   * @return A new, mutable, CommandHandlerBuilderByState
+   */
+  def forNullState(): CommandHandlerBuilderByState[Command, Event, State, State] = {
+    val predicate: Predicate[State] = asJavaPredicate(s ⇒ Objects.isNull(s))
+    val builder = CommandHandlerBuilderByState.builder[Command, Event, State](predicate)
+    builders = builder :: builders
+    builder
+  }
+
+  /**
+   * The handlers defined by this builder are used for any not `null` state.
+   *
+   * @return A new, mutable, CommandHandlerBuilderByState
+   */
+  def forNonNullState(): CommandHandlerBuilderByState[Command, Event, State, State] = {
+    val predicate: Predicate[State] = asJavaPredicate(s ⇒ Objects.nonNull(s))
+    val builder = CommandHandlerBuilderByState.builder[Command, Event, State](predicate)
+    builders = builder :: builders
+    builder
+  }
+
+  /**
+   * The handlers defined by this builder are used for any state.
+   * This variant is particular useful for models that have a single type (ie: no class hierarchy).
+   *
```

Comment:
Document what happens when you combine this with other `forNnn` builders (will it be the default if no other matches or will it trump others)